File: runtime/typescript/packages/foundry/tests/discovery-vectors.test.ts
/**
 * Model-discovery vector tests — validate the Foundry / Azure OpenAI wire →
 * `ModelInfo` mapping against the shared spec vectors in
 * `spec/vectors/discovery/`.
 *
 * Ported from the Rust reference suite
 * (runtime/rust/prompty-foundry/tests/discovery_vectors.rs) so the contract is
 * executed against the TypeScript emitted models rather than assumed. The
 * `shape` field disambiguates the two Foundry endpoints: `deployment` entries
 * go through `deploymentToModelInfo`, `catalog` entries through
 * `catalogModelToModelInfo`. This test only maps (no network).
 */

import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { describe, expect, it } from "vitest";

import { catalogModelToModelInfo, deploymentToModelInfo } from "../src/azure-models.js";

interface DiscoveryVector {
  name: string;
  provider: string;
  shape?: string;
  input: Record<string, unknown>;
  expected: Record<string, unknown>;
}

const vectorsPath = resolve(
  import.meta.dirname,
  "../../../../../spec/vectors/discovery/discovery_vectors.json",
);

const vectors = (
  JSON.parse(readFileSync(vectorsPath, "utf8")) as { vectors: DiscoveryVector[] }
).vectors.filter((v) => v.provider === "foundry");

describe("Foundry discovery vectors", () => {
  it("exercises at least one vector", () => {
    expect(vectors.length).toBeGreaterThan(0);
  });

  for (const vector of vectors) {
    it(`maps ${vector.name} to the canonical ModelInfo shape`, () => {
      const map = vector.shape === "catalog" ? catalogModelToModelInfo : deploymentToModelInfo;
      const actual = map(vector.input).save();
      expect(actual).toEqual(vector.expected);
    });
  }
});

File: runtime/java/prompty-foundry/src/test/java/com/microsoft/prompty/foundry/DiscoveryVectorsTest.java
package com.microsoft.prompty.foundry;

import com.microsoft.prompty.SpecVectors;
import com.microsoft.prompty.model.ModelInfo;
import com.microsoft.prompty.model.SaveContext;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;

/**
 * Grades the Foundry half of the shared discovery suite.
 *
 * <p>Foundry answers from two endpoints, so each vector names the shape it carries and is routed to
 * the matching mapper.
 */
class DiscoveryVectorsTest {

  @TestFactory
  Iterable<DynamicTest> discoveryVectors() {
    List<DynamicTest> tests = new ArrayList<>();
    for (Map<String, Object> vector : SpecVectors.readCases("discovery/discovery_vectors.json", "vectors")) {
      if (!"foundry".equals(vector.get("provider"))) {
        continue;
      }
      String name = SpecVectors.string(vector, "name");
      String shape = SpecVectors.string(vector, "shape");
      tests.add(
          DynamicTest.dynamicTest(
              name,
              () -> {
                Map<String, Object> input = SpecVectors.map(vector, "input");
                ModelInfo actual =
                    switch (shape) {
                      case "deployment" -> FoundryModels.deploymentToModelInfo(input);
                      case "catalog" -> FoundryModels.catalogModelToModelInfo(input);
                      default -> throw new AssertionError("Unknown shape: " + shape);
                    };
                SpecVectors.assertEquivalent(
                    name, vector.get("expected"), actual.save(new SaveContext()));
              }));
    }
    return tests;
  }
}
